### PR TITLE
Introduce a reentrant-mutex-like logic for suppress_instrumentation

### PR DESCRIPTION
**Pull request by reyang**
Trying to help to unblock https://github.com/open-telemetry/opentelemetry-dotnet/pull/1184#discussion_r479441268.

### CODE REVIEW

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs [hunk 1]
```diff
@@ -68,7 +68,12 @@ static TracerProviderSdk()
                 // Callback when Activity is started.
                 ActivityStarted = (activity) =>
                 {
-                    if (!Sdk.SuppressInstrumentation && activity.IsAllDataRequested)
+                    if (!activity.IsAllDataRequested)
+                    {
+                        return;
+                    }
+
+                    if (SuppressInstrumentationScope.TryEnter() == 0)
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:76]: Should be `if (SuppressInstrumentationScope.TryEnter() <= 0)` no?

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs [hunk 2]
```diff
@@ -78,7 +83,12 @@ static TracerProviderSdk()
                 // Callback when Activity is stopped.
                 ActivityStopped = (activity) =>
                 {
-                    if (!Sdk.SuppressInstrumentation && activity.IsAllDataRequested)
+                    if (!activity.IsAllDataRequested)
+                    {
+                        return;
+                    }
+
+                    if (SuppressInstrumentationScope.TryLeave() == 0)
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:91]: Should be `if (SuppressInstrumentationScope.TryLeave() <= 0)` no?

**Comment by CodeBlanch** [comment]
There has to be _some_ feedback about names, right?
```csharp
                ActivityStarted = (activity) =>
                {
                    if (SuppressInstrumentationScope.TryEnter() == 0)
                    {
```

That `TryEnter` looked very strange when I first looked at the code. I was thinking, wait, why are we triggering suppression automatically here?! What that really does is "increment-if-triggered-else-nothing", meaning, it doesn't always enter. Which is why the Try is there, but it's a bit confusing given the behavior of [TryEnter](https://docs.microsoft.com/en-us/dotnet/api/system.threading.monitor.tryenter?view=netcore-3.1) on `Monitor`. This has a different behavior so I feel like we should name it something else. I try to never ask for a rename without a suggestion so here goes... `IncrementIfTriggered`? 🤷

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs (hunk 1 above)
Comment (reyang) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: It is intentional, I've added some comment to clarify the logic.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs (hunk 2 above)
Comment (reyang) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: It is intentional, I've added some comment to clarify the logic.

**Comment by reyang** [comment]
> That `TryEnter` looked very strange when I first looked at the code. I was thinking, wait, why are we triggering suppression automatically here?! What that really does is "increment-if-triggered-else-nothing", meaning, it doesn't always enter. Which is why the Try is there, but it's a bit confusing given the behavior of [TryEnter](https://docs.microsoft.com/en-us/dotnet/api/system.threading.monitor.tryenter?view=netcore-3.1) on `Monitor`. This has a different behavior so I feel like we should name it something else. I try to never ask for a rename without a suggestion so here goes... `IncrementIfTriggered`? 🤷

Taken the suggested name and also changed the `TryLeave` to `DecrementIfTriggered`. It is still a bit weird but much better than the previous one! 😄

**Comment by ejsmith** [comment]
I tried to use this in the Elasticsearch instrumentation provider I'm working on and it didn't seem to work. I called it like this:
```csharp
this.activitySource.Start(activity, ActivityKind.Client);
SuppressInstrumentationScope.Enter();
```
But it didn't seem to end up recording any activities including the one I was trying to record.

**Comment by ejsmith** [comment]
Also, it sure seems like this PR should have included a test to verify that it's working as intended and also to show usage for someone like me trying to figure it out.

**Comment by reyang** [comment]
> I tried to use this in the Elasticsearch instrumentation provider I'm working on and it didn't seem to work. I called it like this:
> 
> ```cs
> this.activitySource.Start(activity, ActivityKind.Client);
> SuppressInstrumentationScope.Enter();
> ```
> 
> But it didn't seem to end up recording any activities including the one I was trying to record.

Follow this PR and see if you could get it work #1184.

**Pull request merged**